### PR TITLE
hdfs: namenode component check

### LINKED ISSUE #241: HDFS Namenode component check

### PR DESCRIPTION
**Pull request by kpgtek**
#### Which issue(s) this PR fixes:

Fixes #241

#### Additional comments:

To make clear that you license your contribution under
the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0) and that you give permission to [TOSIT](https://www.tosit.io/),
you have to acknowledge this by using the following check-box.

 - [x] I hereby declare this contribution to be licenced under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)

 - [x] I hereby agree to grant [TOSIT](https://www.tosit.io/) a copyright license to use my contributions.

### CODE REVIEW

File: tdp_vars_defaults/hadoop/hadoop.yml [hunk 1]
```diff
@@ -166,3 +166,6 @@ jmx_exporter:
   lowercaseOutputName: false
   lowercaseOutputLabelNames: false
 
+# HDFS component listen port
```
Comment (rpignolet) [inline tdp_vars_defaults/hadoop/hadoop.yml:169]:
```suggestion
# HDFS Namenode component listen port
```

File: tdp_vars_defaults/hadoop/hadoop.yml [hunk 2]
```diff
@@ -166,3 +166,6 @@ jmx_exporter:
...
+hdfs_nn_rpc_port: 8020
```
Comment (leopaul36) [inline tdp_vars_defaults/hadoop/hadoop.yml:170]: Shouldn't we parse this from the `dfs.namenode.rpc-address.mycluster.nn` variable?

Comment (kpgtek) [reply tdp_vars_defaults/hadoop/hadoop.yml:170]: We could parse the port number from ```dfs.namenode.rpc-address.mycluster.nn```.
Or could we move ```hdfs_nn_rpc_port:8020``` from ```hadoop.yml``` to ```hdfs.yml```, for futur use (by other service)?

Comment (gboutry) [reply tdp_vars_defaults/hadoop/hadoop.yml:170]: I don't think we should parse these ports, I think it's better to define them as variables in hadoop, it gives us access to these ports easily in other configurations, such as for `knox` where we need to put the ports in conf

Comment (rpignolet) [reply tdp_vars_defaults/hadoop/hadoop.yml:170]: I think we should only put port variable inside `hadoop.yml` when we need it in multiple services. I do not think that Knox need the Namenode RPC port. If a service need it in the futur, we can move the variable from `hdfs.yml` to `hadoop.yml`.

Comment (kpgtek) [reply tdp_vars_defaults/hadoop/hadoop.yml:170]: I think it is more interesting to define explicitly ports variables, if we decide to perform checks on every ports in ```component check```. For the moment, we could "prepare" this by defining them inside each compoment variable file, if needed, we could export them to ```hadoop.yml``` as it is done for ```zookeeper``` (PR #255).

**Comment by leopaul36** [comment]
I have an issue with a fresh install with this refactor. Datanodes are unable to start, the error is:

```
2022-06-03 09:27:19,356 ERROR org.apache.hadoop.hdfs.server.datanode.DataNode: Exception in secureMain
java.net.BindException: Problem binding to [0.0.0.0:9867] java.net.BindException: Address already in use; For more details see:  http://wiki.apache.org/hadoop/BindException
        at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
        at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
        at org.apache.hadoop.net.NetUtils.wrapWithMessage(NetUtils.java:831)
        at org.apache.hadoop.net.NetUtils.wrapException(NetUtils.java:736)
        at org.apache.hadoop.ipc.Server.bind(Server.java:562)
        at org.apache.hadoop.ipc.Server$Listener.<init>(Server.java:1038)
        at org.apache.hadoop.ipc.Server.<init>(Server.java:2810)
        at org.apache.hadoop.ipc.RPC$Server.<init>(RPC.java:960)
        at org.apache.hadoop.ipc.ProtobufRpcEngine$Server.<init>(ProtobufRpcEngine.java:421)
        at org.apache.hadoop.ipc.ProtobufRpcEngine.getServer(ProtobufRpcEngine.java:342)
        at org.apache.hadoop.ipc.RPC$Builder.build(RPC.java:802)
        at org.apache.hadoop.hdfs.server.datanode.DataNode.initIpcServer(DataNode.java:1004)
        at org.apache.hadoop.hdfs.server.datanode.DataNode.startDataNode(DataNode.java:1416)
        at org.apache.hadoop.hdfs.server.datanode.DataNode.<init>(DataNode.java:497)
        at org.apache.hadoop.hdfs.server.datanode.DataNode.makeInstance(DataNode.java:2769)
        at org.apache.hadoop.hdfs.server.datanode.DataNode.instantiateDataNode(DataNode.java:2677)
        at org.apache.hadoop.hdfs.server.datanode.DataNode.createDataNode(DataNode.java:2719)
        at org.apache.hadoop.hdfs.server.datanode.DataNode.secureMain(DataNode.java:2863)
        at org.apache.hadoop.hdfs.server.datanode.DataNode.main(DataNode.java:2887)
Caused by: java.net.BindException: Address already in use
```

I don't really understand this error as i don't see any other process on this port.

Anyway, the default port for `dfs.datanode.address` is 9866 so why change it ?

I don't have the issue if I manually switch back to this with `hdfs_dn_rpc_port: 9866`.

**Comment by leopaul36** [comment]
Ok actually port 9867 is the default for `dfs.datanode.ipc.address` so this is why we have this error.

File: tdp_vars_defaults/hdfs/hdfs.yml [hunk 3]
```diff
@@ -8,6 +8,15 @@ hadoop_hdfs_dir: /var/lib/hdfs
 hadoop_hdfs_pid_dir: /run/hadoop/hdfs
 hadoop_hdfs_log_dir: /var/log/hadoop/hdfs
 
+# HDFS component listen port
+hdfs_nn_rpc_port: 8020
+hdfs_nn_http_port: 9870
+hdfs_nn_https_port: 9871
+hdfs_dn_rpc_port: 9867
```
Comment (leopaul36) [inline tdp_vars_defaults/hdfs/hdfs.yml:15]:
```suggestion
hdfs_dn_rpc_port: 9866
```

File: tdp_vars_defaults/hdfs/hdfs.yml [hunk 4]
```diff
@@ -42,15 +42,15 @@ hdfs_site:
...
   dfs.journalnode.kerberos.principal: jn/_HOST@{{ realm }}
   dfs.journalnode.keytab.file: /etc/security/keytabs/jn.service.keytab
   dfs.journalnode.kerberos.internal.spnego.principal: "HTTP/_HOST@{{ realm }}"
   dfs.block.access.token.enable: "true"
-  dfs.datanode.address: 0.0.0.0:9866
-  dfs.datanode.https.address: 0.0.0.0:9865
-  dfs.datanode.data.dir: "/data/hdfs/dn"
+  dfs.datanode.address: 0.0.0.0:{{ hdfs_dn_rpc_port | string }}
+  dfs.datanode.https.address: 0.0.0.0:{{ hdfs_dn_https_port | string }}
+  dfs.datanode.data.dir: "{{ hadoop_hdfs_dir }}/dn"
```
Comment (rpignolet) [inline tdp_vars_defaults/hdfs/hdfs.yml:53]: You should not modified this variable, we said that data must be in `/data` and `hadoop_hdfs_dir` are for HDFS working files but not HDFS data files.

Comment (kpgtek) [reply tdp_vars_defaults/hdfs/hdfs.yml]: This is fixed, I should make some confusion when beginning this PR for HDFS data files.

**Review by rpignolet — APPROVED** [review]

**Review by rpignolet — CHANGES_REQUESTED** [review]
@kpgtek please rebase on master due to #495.

**Review by gboutry — APPROVED** [review]

**Review by rpignolet — APPROVED** [review]

**Review by leopaul36 — APPROVED** [review]

**Pull request merged**